src/main/java/org/truffleruby/language/CallStackManager.java

```diff
@@ -185,43 +185,58 @@ public void printBacktrace(Node currentNode) {
+        return getBacktraceForException(currentNode, 0, exceptionClass, null);
     }
 
-    public Backtrace getBacktrace(Node currentNode) {
-        return getBacktrace(currentNode, 0, false, null, null);
+    public Backtrace getBacktraceForException(Node currentNode, int omit, DynamicObject exceptionClass) {
+        return getBacktraceForException(currentNode, omit, exceptionClass, null);
     }
 
-    public Backtrace getBacktrace(Node currentNode, int omit) {
-        return getBacktrace(currentNode, omit, false, null, null);
+    public Backtrace getBacktraceForException(Node currentNode, DynamicObject exceptionClass, Throwable javaThrowable) {
+        return getBacktraceForException(currentNode, 0, exceptionClass, javaThrowable);
     }
 
-    public Backtrace getBacktrace(Node currentNode, int omit, DynamicObject exception) {
-        return getBacktrace(currentNode, omit, false, exception, null);
+    @TruffleBoundary
+    public Backtrace getBacktraceForException(Node currentNode,
+                                  final int omit,
+                                  DynamicObject exceptionClass,
+                                  Throwable javaThrowable) {
+        assert exceptionClass != null;
+        if (context.getOptions().BACKTRACES_OMIT_UNUSED
+                && DisablingBacktracesNode.areBacktracesDisabled()
+                && ModuleOperations.assignableTo(exceptionClass, context.getCoreLibrary().getStandardErrorClass())) {
+            return new Backtrace(currentNode, new Activation[] { Activation.OMITTED_UNUSED }, null);
+        }
+        return getBacktrace(currentNode, omit, false, javaThrowable);
+    }
+
+    public Backtrace getBacktrace(Node currentNode) {
+        return getBacktrace(currentNode, 0, false, null);
     }
 
     public Backtrace getBacktrace(Node currentNode,
-                                  final int omit,
-                                  final boolean filterNullSourceSection,
-                                  DynamicObject exception) {
-        return getBacktrace(currentNode, omit, filterNullSourceSection, exception, null);
+            final int omit,
+            final boolean filterNullSourceSection) {
+        return getBacktrace(currentNode, omit, filterNullSourceSection, null);
     }
 
     @TruffleBoundary
-    public Backtrace getBacktrace(Node currentNode,
-                                  final int omit,
-                                  final boolean filterNullSourceSection,
-                                  DynamicObject exception,
-                                  Throwable javaThrowable) {
-        if (exception != null
-                && context.getOptions().BACKTRACES_OMIT_UNUSED
-                && DisablingBacktracesNode.areBacktracesDisabled()
-                && ModuleOperations.assignableTo(Layouts.BASIC_OBJECT.getLogicalClass(exception),
-                    context.getCoreLibrary().getStandardErrorClass())) {
```

Comment:
So the problem was that the exception was not passed in many cases, and we often don't have an exception instance at this point yet, so now we pass the exception class consistently to perform this check?

Reply:
There were several very distinct cases for filling in backtraces.
1. When raising an exception in ruby we have the exception and may need to fill in the stack trace
2. When raising an exception from Java we typically had not created the exception object yet.
3. Other stack related operations such as looking for callers, or creating a backtrace, etc.

I hoped to fix this with a fairly minimal change, but everywhere that created exceptions built its own backtrace, so I refactored things to reduce the number of places we did that and make it easier to sort things out.

Reply:
Thanks for the explanation!